File: agent/internal/noise/handshake_test.go
// agent/internal/noise/handshake_test.go
package noise

import (
	"bytes"
	"testing"
)

// TestFullHandshake runs a complete Noise_XX handshake between two in-process
// peers and verifies that post-handshake encryption works in both directions.
func TestFullHandshake(t *testing.T) {
	// Responder (agent) has a static key pair
	responderStatic, err := generateKeypair()
	if err != nil {
		t.Fatalf("generate responder static: %v", err)
	}

	initiator, err := NewInitiator()
	if err != nil {
		t.Fatalf("NewInitiator: %v", err)
	}
	responder, err := NewResponder(responderStatic)
	if err != nil {
		t.Fatalf("NewResponder: %v", err)
	}

	// Message 1: initiator -> responder (65 bytes)
	msg1, err := initiator.WriteMessage1()
	if err != nil {
		t.Fatalf("WriteMessage1: %v", err)
	}
	if len(msg1) != 65 {
		t.Fatalf("msg1 length: got %d, want 65", len(msg1))
	}
	if err := responder.ReadMessage1(msg1); err != nil {
		t.Fatalf("ReadMessage1: %v", err)
	}

	// Message 2: responder -> initiator (162 bytes)
	msg2, err := responder.WriteMessage2()
	if err != nil {
		t.Fatalf("WriteMessage2: %v", err)
	}
	if len(msg2) != 162 {
		t.Fatalf("msg2 length: got %d, want 162", len(msg2))
	}
	if err := initiator.ReadMessage2(msg2); err != nil {
		t.Fatalf("ReadMessage2: %v", err)
	}

	// Initiator must see responder's static public key after msg2
	if !bytes.Equal(initiator.RemoteStaticPub(), responderStatic.PublicKey().Bytes()) {
		t.Fatal("initiator RemoteStaticPub mismatch after ReadMessage2")
	}

	// Message 3: initiator -> responder (97 bytes)
	msg3, err := initiator.WriteMessage3()
	if err != nil {
		t.Fatalf("WriteMessage3: %v", err)
	}
	if len(msg3) != 97 {
		t.Fatalf("msg3 length: got %d, want 97", len(msg3))
	}
	if err := responder.ReadMessage3(msg3); err != nil {
		t.Fatalf("ReadMessage3: %v", err)
	}

	// Both sides split into cipher states
	iSend, iRecv := initiator.Split()
	rSend, rRecv := responder.Split()

	// Initiator sends to responder
	plaintext := []byte("hello from browser")
	ct, err := iSend.encryptWithAd(nil, plaintext)
	if err != nil {
		t.Fatalf("iSend.encrypt: %v", err)
	}
	pt, err := rRecv.decryptWithAd(nil, ct)
	if err != nil {
		t.Fatalf("rRecv.decrypt: %v", err)
	}
	if !bytes.Equal(pt, plaintext) {
		t.Fatalf("i->r plaintext mismatch: got %q, want %q", pt, plaintext)
	}

	// Responder sends to initiator
	reply := []byte("hello from agent")
	ct2, err := rSend.encryptWithAd(nil, reply)
	if err != nil {
		t.Fatalf("rSend.encrypt: %v", err)
	}
	pt2, err := iRecv.decryptWithAd(nil, ct2)
	if err != nil {
		t.Fatalf("iRecv.decrypt: %v", err)
	}
	if !bytes.Equal(pt2, reply) {
		t.Fatalf("r->i plaintext mismatch: got %q, want %q", pt2, reply)
	}
}

func TestHandshakeRejectsOutOfOrderCalls(t *testing.T) {
	initiator, err := NewInitiator()
	if err != nil {
		t.Fatalf("NewInitiator: %v", err)
	}
	if _, err := initiator.WriteMessage3(); err == nil {
		t.Fatal("WriteMessage3 before ReadMessage2 must fail")
	}

	responderStatic, err := generateKeypair()
	if err != nil {
		t.Fatalf("generate responder static: %v", err)
	}
	responder, err := NewResponder(responderStatic)
	if err != nil {
		t.Fatalf("NewResponder: %v", err)
	}
	if _, err := responder.WriteMessage2(); err == nil {
		t.Fatal("WriteMessage2 before ReadMessage1 must fail")
	}
}

File: signaling-server/web/noise-p256/handshake.test.js
// signaling-server/web/noise-p256/handshake.test.js
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { NoiseXX } from './index.js'

test('full Noise_XX handshake: mutual encryption works in both directions', async () => {
  const initiator = await NoiseXX.createInitiator()
  const responderStatic = await (await import('./keys.js')).generateKeypair()
  const responder = await NoiseXX.createResponder(responderStatic.privateKey, responderStatic.publicKeyBytes)

  // Message 1: 65 bytes
  const msg1 = await initiator.writeMessage1()
  assert.strictEqual(msg1.length, 65)
  await responder.readMessage1(msg1)

  // Message 2: 162 bytes
  const msg2 = await responder.writeMessage2()
  assert.strictEqual(msg2.length, 162)
  await initiator.readMessage2(msg2)

  // Initiator must see responder's static pub after msg2
  assert.deepStrictEqual(initiator.remoteStaticPub, responderStatic.publicKeyBytes)

  // Message 3: 97 bytes
  const msg3 = await initiator.writeMessage3()
  assert.strictEqual(msg3.length, 97)
  await responder.readMessage3(msg3)

  // Split into cipher states
  const [iSend, iRecv] = await initiator.split()
  const [rSend, rRecv] = await responder.split()

  const plaintext = new TextEncoder().encode('hello from browser')
  const ct = await iSend.encrypt(new Uint8Array(0), plaintext)
  const pt = await rRecv.decrypt(new Uint8Array(0), ct)
  assert.deepStrictEqual(pt, plaintext)

  const reply = new TextEncoder().encode('hello from agent')
  const ct2 = await rSend.encrypt(new Uint8Array(0), reply)
  const pt2 = await iRecv.decrypt(new Uint8Array(0), ct2)
  assert.deepStrictEqual(pt2, reply)
})

test('message lengths match spec', async () => {
  const initiator = await NoiseXX.createInitiator()
  const rs = await (await import('./keys.js')).generateKeypair()
  const responder = await NoiseXX.createResponder(rs.privateKey, rs.publicKeyBytes)

  const msg1 = await initiator.writeMessage1()
  await responder.readMessage1(msg1)
  const msg2 = await responder.writeMessage2()
  await initiator.readMessage2(msg2)
  const msg3 = await initiator.writeMessage3()

  assert.strictEqual(msg1.length, 65,  'msg1 must be 65 bytes')
  assert.strictEqual(msg2.length, 162, 'msg2 must be 162 bytes')
  assert.strictEqual(msg3.length, 97,  'msg3 must be 97 bytes')
})

test('out-of-order calls are rejected', async () => {
  const initiator = await NoiseXX.createInitiator()
  await assert.rejects(() => initiator.writeMessage3())

  const rs = await (await import('./keys.js')).generateKeypair()
  const responder = await NoiseXX.createResponder(rs.privateKey, rs.publicKeyBytes)
  await assert.rejects(() => responder.writeMessage2())
})
